klaus/contrib/auto_klaus.py

```diff
@@ -0,0 +1,103 @@
+"""
+Alternative take on the "automatically discovered repositories" concept
+that requires no threads, polling or inotify. Instead the filesystem is
+consulted whenever a repository name is looked up.
+
+Since os.path.exists() and os.listdir() are fairly quick filesystem
+operations, performance should be good for small to medium sites.
+FancyRepo() objects are cached.
+
+Repositories are identified by the existence of a
+
+    <reponame>/git-daemon-export-ok
+
+file (for compatibility with gitweb).
+
+Example usage:
+
+    from klaus.contrib.auto_klaus import Klaus, SlashDynamicRepos
+
+    application = Klaus('/srv/git', "My git repositories", False)
+
+    application.wsgi_app = httpauth.AlwaysFailingAuthMiddleware(
+        wsgi_app=dulwich.web.make_wsgi_chain(
+            backend=dulwich.server.DictBackend(SlashDynamicRepos(application.valid_repos)),
+            fallback_app=application.wsgi_app,
+        ),
+    )
+"""
+
+import collections.abc
+import pathlib
+import os
+import os.path
+
+import klaus
+import klaus.repo
+
+class FilesystemRepoDict(collections.abc.Mapping):
+    """
+    Maintain a virtual read-only dictionary whose contents represent
+    the presence of git repositories in the given root directory.
+    """
+
+    def __init__(self, root):
+        self._root = pathlib.Path(root)
+        self._repos = {}
+
+    def __getitem__(self, name):
+        if not name or name[0] == '.' or '/' in name or '\0' in name:
```

Comment:
I think we should also check for `os.pardir` here?

Reply:
It's included in `name[0] == '.'` for most common operating systems but we could include `os.pardir` and `os.curdir` explicitly, yes.

Reply:
Added all of these to the validation expression.